MLPipelines/TrainDatasets.py

```diff
@@ -229,44 +232,65 @@ def fetch_image_partitions(self) -> pd.DataFrame:
         })
         return df.drop_duplicates()
 
-    def image_sampler(self, project_ids: List[int], skip_tags: List[str] = [], image_feather: str = None, anno_feather: str = None, p: float = 0.2) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    def image_sampler(self, project_ids: List[int], skip_tags: List[str] = [], image_feather: str = None, anno_feather: str = None, test_split: bool = False, p: float = 0.2) -> Union[Tuple[pd.DataFrame, pd.DataFrame], Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]]:
         train_df = []
         validate_df = []
-        image_names = self.unique_image_names(project_ids, skip_tags, image_feather)
-        print('Found', len(image_names), 'unique images')
-        np.random.shuffle(image_names)
-        val_images = image_names[:int(len(image_names)*p)]
-        anno_df = []
+        test_df = []
+
+        # Get annotations from disk or datalake
         if anno_feather is not None:
             anno_df = []
             for project_id in project_ids:
                 anno_df.append(feather.read_feather(anno_feather.format(project_id)))
             anno_dfs = pd.concat(anno_df)
-        for image_name in image_names:
-            if anno_feather is not None:
-                anno_df = anno_dfs[anno_dfs['image_name'] == image_name]
-            else:
-                anno_df = self.annotations.to_table(columns=self.anno_proj, filter=ds.field("image_name") == image_name).to_pandas()
-            if image_name in val_images:
-                print("Adding", image_name, "to val")
+        else:
+            self.create_datasources()
+            print('Filtering annotations')
+            anno_dfs = self.annotations.to_table(columns=self.anno_proj, filter=(ds.field("project_id").isin(project_ids))).to_pandas()
+
+        # Get image names for project
+        image_names = self.unique_image_names(project_ids, skip_tags, image_feather)
+        image_names = pd.DataFrame({ 'image_name': image_names })
+        print(f"Found {len(image_names)} unique images")
+        
+        # Split into train, validation, and test
+        if test_split:
+            split_images = image_names.sample(frac=p)
+            val_images = split_images.sample(frac=0.8)
```

Comment:
frac here should be (1.0 - p)

Reply:
split_images is a sample of p% of total images, then val and test are 80% and 20% of split. In the loop on line 265 the dataframes are appended to arrays depending on if the image name appears in val_images, test_images, or else (image_names).